# 🚀 Fine-Tuning Gemma 3 270M (or Higher) on CPU/GPU with LoRA

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1DUnWNxLMxL3WprUqhyLayPxVr5KffTsC?usp=sharing)


## 📖 About This Tutorial

### 👥 Who Is This For?

This comprehensive guide is designed for **everyone** interested in fine-tuning language models:

- 🎓 **Students & Researchers**: Learn the complete workflow from data creation to model deployment
- 💼 **ML Engineers**: Understand practical implementation details and best practices
- 🚀 **AI Enthusiasts**: Get hands-on experience with state-of-the-art fine-tuning techniques
- 🏢 **Domain Experts**: Adapt language models to your specific field (medical, legal, scientific, etc.)

**No prior fine-tuning experience required!** We'll walk through everything step-by-step.

---

### 🎯 What You'll Learn

In this tutorial, we'll fine-tune **Gemma 3 270M** on a custom radiobiology dataset. You'll master:

1. 📊 **Data Creation**: Generate high-quality training data from scratch using LLM-DATA-Generator
2. 🔧 **Model Setup**: Load and configure Gemma 3 models with KerasHub
3. 🎛️ **LoRA Fine-Tuning**: Apply parameter-efficient fine-tuning techniques
4. 📈 **Training**: Optimize hyperparameters and monitor training progress
5. 🧪 **Evaluation**: Test model performance before and after fine-tuning
6. 💾 **Deployment**: Save and use your fine-tuned model for inference

---

### 💻 Hardware Requirements: CPU-First Approach

#### 🖥️ This Tutorial Uses CPU

We'll be fine-tuning **Gemma 3 270M on CPU** to make this tutorial **accessible to everyone**!

✅ **No expensive GPU required**  
✅ **Runs on any laptop or desktop**  
✅ **Perfect for learning and experimentation**  
✅ **Same code works everywhere**

---

#### 🚀 Want Faster Training? Use the Same Code!

The beauty of this tutorial is that **the exact same code** can run on different hardware:

**Option 1: CPU (What We're Using)** 🖥️
- ✅ Works on any device
- ✅ No special setup required
- ✅ Great for learning
- ⏱️ Training time: Slower but totally doable for Gemma 270M

**Option 2: GPU (Same Code, Faster!)** ⚡
- ✅ Just install GPU-enabled packages:
  ```bash
  pip install keras-hub[jax] jax[cuda12]  # For NVIDIA GPUs
  ```
- ✅ **No code changes needed** - it automatically detects your GPU!
- ⚡ Training time: Much faster (minutes instead of hours)
- 💡 Recommended for larger models or frequent fine-tuning

**Option 3: Google Colab (Free GPU!)** 🌟
- ✅ Upload this notebook to [Google Colab](https://colab.research.google.com/)
- ✅ Runtime → Change runtime type → GPU (T4 with 15GB VRAM)
- ✅ **Same code, zero installation, free GPU access!**
- 💡 Perfect if you want speed without buying hardware

---


### 🗺️ The Complete Workflow

Here's what we'll accomplish in this tutorial:

```
📚 Step 1: Data Creation
   ├─ Source: Radiobiology textbook (PDF)
   ├─ Tool: LLM-DATA-Generator
   └─ Output: 832 Q&A pairs (CSV)
   
⬇️

🔧 Step 2: Environment Setup
   ├─ Install Keras & KerasHub
   ├─ Configure Kaggle credentials
   └─ Set JAX backend for optimal performance
   
⬇️

🤖 Step 3: Model Loading
   ├─ Download Gemma 3 270M from Kaggle
   ├─ Load pre-trained weights
   └─ Test baseline performance
   
⬇️

🎯 Step 4: LoRA Configuration
   ├─ Enable LoRA (rank=8)
   ├─ Freeze base model weights
   └─ Configure trainable parameters
   
⬇️

📈 Step 5: Training
   ├─ Prepare dataset
   ├─ Set hyperparameters
   └─ Train the model
   
⬇️

✅ Step 6: Evaluation & Inference
   ├─ Compare before/after performance
   ├─ Test on domain-specific questions
   └─ Save fine-tuned model
```

---

### ⏱️ Time Commitment

- **Reading & Understanding**: ~30 minutes
- **Running the Code** (with GPU): ~45-60 minutes
- **Total**: ~1.5-2 hours for complete mastery

---

### 📦 What's Included

- ✅ Complete, runnable code with detailed explanations
- ✅ Pre-generated dataset (832 Q&A pairs)
- ✅ Step-by-step instructions with visual aids
- ✅ Best practices and optimization tips
- ✅ Troubleshooting guidance
- ✅ Links to additional resources

---

### 🚀 Ready to Get Started?

Let's dive into the fine-tuning architecture and understand how all the pieces fit together!

---



## 🏗️ The Fine-Tuning Architecture

### 🧠 Pop Quiz #3: Understanding the Flow

**Question**: In what order should we fine-tune a model?

- A) Train first, then worry about data later 🤪
- B) Get data → Load model → Train → Evaluate 🎯
- C) Just hope for the best 🙏
- D) Download model → Pray to the AI gods 🕯️

<details>
<summary><b>👉 Click for the answer!</b></summary>

**Answer: B!** Let's see the architecture:

</details>

![Fine-Tuning Architecture](Architecture.png)

### Understanding the Architecture 🎨

The diagram above shows our complete fine-tuning pipeline.
   

## 📊 Step 1: Data Generation - Creating Your Training Dataset

### The Challenge 🤔

**Quick question**: How long would it take YOU to manually create 1,000 high-quality Q&A pairs from a textbook?

- A) A few hours ⏰
- B) A few days 📅
- C) A few weeks 😰
- D) I'd rather not... 💀

**Answer**: Probably C or D! That's why we automate it. 🤖

### Solution: LLM-DATA-Generator ⚡

We use the **LLM-DATA-Generator** tool to automatically generate training data from the Radiobiology textbook.

**🔗 Tool Repository**: [ElHadheqMind/LLM-DATA-Generator](https://github.com/ElHadheqMind/LLM-DATA-Generator)

### How It Works (The Simple Version)

```
📖 Input: Radiobiology Textbook (PDF)
    ↓
🤖 LLM-DATA-Generator Magic
    ↓
📊 Output: 832 Q&A Pairs (CSV)
    ↓
✅ Ready for Fine-Tuning!
```

**Time saved**: ~2 weeks of manual work → 30 minutes automated! 🎉

---

### 🎬 Dataset Generation Process in Action

Watch this step-by-step walkthrough showing how to generate the dataset using LLM-DATA-Generator:

![Dataset Generation Demo](llm-data-generator-demo.gif)

**What the video covers**:
1. ⚙️ **Setup**: Installing and configuring LLM-DATA-Generator
2. 🔑 **API Configuration**: Setting up LLM API credentials (OpenAI/Anthropic/Google)
3. 📄 **Document Upload**: Loading the source textbook (PDF/DOCX)
4. 🎛️ **Parameter Tuning**: Configuring generation settings (question count, difficulty, format)
5. 🚀 **Generation**: Running the automated Q&A extraction process
6. ✅ **Export**: Reviewing and exporting the final CSV dataset

**🔗 Tool Repository**: [ElHadheqMind/LLM-DATA-Generator](https://github.com/ElHadheqMind/LLM-DATA-Generator)

---

### Dataset Structure

Our generated dataset contains:

| Column | Description | Example |
|--------|-------------|----------|
| **Question** | Domain-specific question | "What is the primary mechanism by which radiation damages DNA?" |
| **Answer** | Comprehensive answer with context | "Radiation damages DNA primarily through..." |
| **Content** | Source text from the textbook | Original section from the book |

---

## Setup

To complete this tutorial, you will first need to complete the setup instructions at [Gemma setup](https://ai.google.dev/gemma/docs/setup). The Gemma setup instructions show you how to do the following:

* Get access to Gemma on [kaggle.com](https://kaggle.com).
* Optional: GPU resources (RTX 4060 8GB or similar) for faster training, but **CPU works fine** for this tutorial with Gemma 3 270M.
  [Learn more](https://ai.google.dev/gemma/docs/core#sizes).
* Generate and configure a Kaggle username and API key.

After you've completed the Gemma setup, move on to the next section, where you'll set environment variables for your local environment.

### 🔑 Kaggle Configuration

Gemma models are hosted on Kaggle. You'll need API credentials to download them programmatically.

---

#### **Step 1: Create Kaggle Account & Accept Terms**

1. **Sign up**: [kaggle.com/account/login](https://www.kaggle.com/account/login?phase=startRegisterTab)
2. **Accept Gemma license**: [kaggle.com/models/google/gemma-3](https://www.kaggle.com/models/google/gemma-3)  
   Click any variant → **"Request Access"** → Accept terms

**📚 Reference**: [Kaggle Models Documentation](https://www.kaggle.com/docs/models)

---

#### **Step 2: Generate API Token**

1. Profile → **Settings** → Scroll to **API** section
2. Click **"Create New Token"** → Downloads `kaggle.json`

**File structure**:
```json
{"username": "your_username", "key": "abc123..."}
```

**🔒 Security**: This file grants full API access. Never commit to Git or share publicly.

**📚 Reference**: [Kaggle API Authentication](https://www.kaggle.com/docs/api#authentication)

---


### 🔐 Load Kaggle Credentials

This cell loads your Kaggle credentials and sets the required environment variables (`KAGGLE_USERNAME` and `KAGGLE_KEY`).

**How it works**: The Kaggle Python library checks these environment variables when downloading models. See [kaggle-api source](https://github.com/Kaggle/kaggle-api/blob/main/kaggle/api_client.py#L53).

In [1]:
import os
import json

# ============================================================================
# Reads credentials from a local kaggle.json file in the current directory.
# This approach keeps credentials out of code and works across environments.
# ============================================================================

with open('kaggle.json', 'r') as f:
    kaggle_creds = json.load(f)

os.environ["KAGGLE_USERNAME"] = kaggle_creds['username']
os.environ["KAGGLE_KEY"] = kaggle_creds['key']

# Optional: Set additional Kaggle environment variables
# os.environ["KAGGLE_PROXY"] = "http://proxy.example.com:8080"  # For corporate proxies
# os.environ["KAGGLE_CONFIG_DIR"] = "/custom/path/.kaggle"      # Override config directory

print("✅ Kaggle credentials loaded from kaggle.json")
print(f"   Username: {kaggle_creds['username']}")


✅ Kaggle credentials loaded from kaggle.json
   Username: mezzihoussem


### 📦 Install Required Packages

We'll install the latest versions of **Keras** and **KerasHub**:

- **Keras 3**: Multi-framework deep learning API (supports JAX, TensorFlow, PyTorch)
- **KerasHub**: Pre-trained models and utilities for NLP tasks, including Gemma

These packages provide:
- Easy access to Gemma models via `from_preset()`
- Built-in LoRA support with `enable_lora()`
- Efficient training utilities and optimizers
- Seamless integration with different backends

In [2]:
# ============================================================================
# PACKAGE INSTALLATION
# ============================================================================
# Install KerasHub (includes pre-trained models) and Keras 3
# The -q flag suppresses verbose output, -U ensures latest versions
# ============================================================================

!pip install -q -U keras-hub
!pip install -q -U keras

print("✅ Packages installed successfully!")

✅ Packages installed successfully!


### ⚙️ Select Keras Backend

**Keras 3** is a multi-framework API that supports three backends:

| Backend | Pros | Cons | Best For |
|---------|------|------|----------|
| **JAX** | Fastest training, XLA compilation, best GPU utilization | Newer ecosystem | Production training |
| **TensorFlow** | Mature ecosystem, TensorFlow Serving | Slower than JAX | Deployment |
| **PyTorch** | Popular, extensive community | Slower than JAX | Research |

**For this tutorial**, we use **JAX** because:
- ✅ Best performance for LoRA fine-tuning
- ✅ Optimized for NVIDIA GPUs (including RTX 4060)
- ✅ Recommended by Google for Gemma models
- ✅ Efficient memory management for 8GB VRAM

**Reference**: [Keras 3 Multi-Backend Documentation](https://keras.io/keras_3/)

In [3]:
# ============================================================================
# BACKEND CONFIGURATION
# ============================================================================
# Set JAX as the Keras backend for optimal performance
# XLA_PYTHON_CLIENT_MEM_FRACTION controls GPU memory allocation (1.00 = 100%)
# This prevents memory fragmentation on JAX backend
# ============================================================================

os.environ["KERAS_BACKEND"] = "jax"  # Options: "jax", "tensorflow", "torch"
os.environ["XLA_PYTHON_CLIENT_MEM_FRACTION"] = "1.00"  # Use full GPU memory

print("✅ Backend configured: JAX")

✅ Backend configured: JAX


### 📚 Import Libraries

Now we'll import the necessary libraries:

- **keras**: Core deep learning framework
- **keras_hub**: Pre-trained models and NLP utilities
- **pandas**: For dataset manipulation
- **numpy**: For numerical operations

In [5]:
# ============================================================================
# LIBRARY IMPORTS
# ============================================================================

import keras
import keras_hub
import pandas as pd
import numpy as np
import sys

print(f"✅ Keras version: {keras.__version__}")
print(f"✅ KerasHub version: {keras_hub.__version__}")
print(f"✅ Backend: {keras.backend.backend()}")

✅ Keras version: 3.12.0
✅ KerasHub version: 0.23.0
✅ Backend: jax


### 📋 Verify Package Versions

Let's verify that all packages are installed with the correct versions. This tutorial was tested with the following configuration:

**Tested Configuration:**
- **Keras**: 3.12.0
- **KerasHub**: 0.23.0
- **TensorFlow**: 2.19.0
- **JAX**: 0.7.2
- **TensorFlow Text**: 2.19.0
- **Python**: 3.12.12

**Note**: Minor version differences are usually fine, but major version mismatches may cause compatibility issues.

In [6]:
# ============================================================================
# VERSION VERIFICATION
# ============================================================================
# Display all package versions for reproducibility and troubleshooting
# ============================================================================

import tensorflow as tf
import jax
try:
    import tensorflow_text as tf_text
    tf_text_version = tf_text.__version__
except ImportError:
    tf_text_version = "Not installed"

print("="*60)
print("PACKAGE VERSIONS")
print("="*60)
print(f"Keras version: {keras.__version__}")
print(f"Keras-Hub version: {keras_hub.__version__}")
print(f"TensorFlow version: {tf.__version__}")
print(f"JAX version: {jax.__version__}")
print(f"TensorFlow Text version: {tf_text_version}")
print(f"Python version: {sys.version}")
print("="*60)
print("\n✅ All packages loaded successfully!")

PACKAGE VERSIONS
Keras version: 3.12.0
Keras-Hub version: 0.23.0
TensorFlow version: 2.19.0
JAX version: 0.7.2
TensorFlow Text version: 2.19.0
Python version: 3.12.3 (main, Aug 14 2025, 17:47:21) [GCC 13.3.0]

✅ All packages loaded successfully!


## 🔽 Load Gemma 3 Model

Now we'll load the pre-trained Gemma 3 model using KerasHub's convenient `from_preset()` method.

### What is a Causal Language Model?

A **Causal Language Model (CLM)** predicts the next token based on previous tokens in a sequence. This is the foundation of text generation:

```
Input:  "The radiation dose is"
Output: "measured in Gray (Gy)"
```

### Available Gemma 3 Presets

KerasHub provides several pre-configured Gemma 3 models:

| Preset Name | Size | Context | Type | Memory |
|-------------|------|---------|------|--------|
| `gemma3_instruct_270m` | 270M | 8K | Text-only | ~0.5 GB |
| `gemma3_instruct_1b` | 1B | 32K | Text-only | ~1.5 GB |
| `gemma3_instruct_4b` | 4B | 128K | Multimodal | ~6.4 GB |
| `gemma3_instruct_12b` | 12B | 128K | Multimodal | ~20 GB |
| `gemma3_instruct_27b` | 27B | 128K | Multimodal | ~46 GB |

**For this tutorial**, we use `gemma3_instruct_270m` - optimized for instruction following and fine-tuning on consumer CPUs and GPUs.

**Browse all models**: [Gemma 3 on Kaggle](https://www.kaggle.com/models/keras/gemma3)

### Loading Process

The `from_preset()` method will:
1. Download the model weights from Kaggle (~1.5 GB)
2. Load the tokenizer (262K vocabulary)
3. Initialize the model architecture
4. Load pre-trained weights

**Note**: First-time download may take 2-5 minutes depending on your connection.

In [8]:
# ============================================================================
# MODEL LOADING
# ============================================================================
# Load Gemma 3 1B Instruct model from Kaggle
# This will download ~1.5 GB on first run (cached for future use)
# ============================================================================

print("📥 Loading Gemma 3 270m Instruct model...")
print("⏳ This may take 2-5 minutes on first run...\n")

gemma_lm = keras_hub.models.Gemma3CausalLM.from_preset("gemma3_instruct_270m")

print("\n✅ Model loaded successfully!")


📥 Loading Gemma 3 270m Instruct model...
⏳ This may take 2-5 minutes on first run...


✅ Model loaded successfully!


The `Gemma3CausalLM.from_preset()` method instantiates the model from a preset architecture and weights. In the code above, the string `"gemma#_xxxxxxx"` specifies a preset version and parameter size for Gemma. You can find the code strings for Gemma models in their **Model Variation** listings on [Kaggle](https://www.kaggle.com/models/keras/gemma3).

## 🔍 Inference Before Fine-Tuning

Before fine-tuning, let's test how Gemma 3 270M responds to different prompt formats. **Gemma 3 270M is highly sensitive to prompt engineering** - the way you structure your prompt significantly impacts response quality.

### 🎯 Sampling Strategy

We use **`keras_hub.samplers.TopPSampler`** ([documentation](https://keras.io/keras_hub/api/samplers/top_p_sampler/)) for controlled text generation:
- **Top-P (nucleus) sampling**: Selects from the smallest set of tokens whose cumulative probability exceeds `p`
- **Top-K filtering**: Limits selection to the `k` most likely tokens
- **Parameters**: `p=0.3, k=5, seed=2` for reproducible, focused outputs

### 📋 Prompt Template Formats

We'll test three approaches:

**1. Custom Template (Simple)**
```python
"System Instruction:\n{system_instruction}\n\nQuestion:\n{instruction}\n\nResponse:\n{response}"
```

**2. Gemma Official Chat Template**
```python
"<start_of_turn>user\n{content}\n<end_of_turn>\n<start_of_turn>model\n"
```
- `<start_of_turn>user` - Marks user input
- `<end_of_turn>` - Signals turn end
- `<start_of_turn>model` - Model response starts here

Let's see how each performs! 🧪

### ❌ Test 1: No Chat Template, No System Instruction

**Approach**: Simple instruction-response format without system context.

**Expected Issue**: Without proper guidance, the model may hallucinate or provide incorrect answers.

In [10]:
template = "Instruction:\n{instruction}\n\nResponse:\n{response}"

prompt = template.format(
    instruction="What is the double-helix molecule primarily responsible for heredity and the synthesis of proteins?",
    response="",
)
sampler = keras_hub.samplers.TopPSampler(p=0.3, k=5, seed=2)
gemma_lm.compile(sampler=sampler)
print(gemma_lm.generate(prompt, max_length=256))

Instruction:
What is the double-helix molecule primarily responsible for heredity and the synthesis of proteins?

Response:
The double-helix molecule primarily responsible for heredity and the synthesis of proteins is a **protein**.
<end_of_turn>


**Result**: ❌ **WRONG!** The model says "protein" instead of "DNA". Without system instruction, Gemma 3 270M struggles with basic questions.

### ✅ Test 2: Custom Template + System Instruction

**Approach**: Add system instruction to guide the model's behavior.

**Hypothesis**: System context should improve accuracy.

In [11]:
SYSTEM_INSTRUCTION = "You are a concise and expert assistant in Radiobiology. Provide accurate, clear, and relevant answers."
instruction_text = "What is the double-helix molecule primarily responsible for heredity and the synthesis of proteins?"
template = "System Instruction:\n{system_instruction}\n\nQuestion:\n{instruction}\n\nResponse:\n{response}"
prompt = template.format(
    system_instruction=SYSTEM_INSTRUCTION,
    instruction=instruction_text,
    response="",
)

sampler = keras_hub.samplers.TopPSampler(p=0.3, k=5, seed=2)
gemma_lm.compile(sampler=sampler)
print(gemma_lm.generate(prompt, max_length=256))

System Instruction:
You are a concise and expert assistant in Radiobiology. Provide accurate, clear, and relevant answers.

Question:
What is the double-helix molecule primarily responsible for heredity and the synthesis of proteins?

Response:
The double-helix molecule primarily responsible for heredity and the synthesis of proteins is the **DNA**.
<end_of_turn>


**Result**: ✅ **CORRECT!** The model now correctly answers "DNA". System instruction provides crucial context.

### ✅ Test 3: Gemma Official Chat Template

**Approach**: Use Gemma's official chat format with turn delimiters.

**Hypothesis**: Official template should work equally well or better.

In [13]:
# --- 1. Define Model-Specific Constants and System Instructions ---
# These are the standard delimiters for Gemma instruction-tuned models.
START_TURN_USER = "<start_of_turn>user\n"
END_TURN = "<end_of_turn>\n"
START_TURN_MODEL = "<start_of_turn>model\n"

# System/Role instruction is prepended to the user's turn
SYSTEM_INSTRUCTION = "You are a concise and expert assistant in Radiobiology. Provide accurate, clear, and relevant answers."
USER_QUESTION = "What is the double-helix molecule primarily responsible for heredity and the synthesis of proteins?"

# Step 2a: Create the content of the user turn
# We wrap the SYSTEM_INSTRUCTION and the USER_QUESTION together.
user_content = f"{SYSTEM_INSTRUCTION}Question:\n{USER_QUESTION}"

# Step 2b: Assemble the full prompt with turn delimiters
# The model is trained to start generating after <start_of_turn>model
prompt = (
    f"{START_TURN_USER}"  # <start_of_turn>user\n
    f"{user_content}\n"
    f"{END_TURN}"         # <end_of_turn>\n
    f"{START_TURN_MODEL}" # <start_of_turn>model\n
)
# --- 3. Configure and Generate ---
sampler = keras_hub.samplers.TopPSampler(p=0.3, k=5, seed=2)

gemma_lm.compile(sampler=sampler)
print(gemma_lm.generate(prompt, max_length=256))

<start_of_turn>user
You are a concise and expert assistant in Radiobiology. Provide accurate, clear, and relevant answers.Question:
What is the double-helix molecule primarily responsible for heredity and the synthesis of proteins?
<end_of_turn>
<start_of_turn>model
The double-helix molecule primarily responsible for heredity and the synthesis of proteins is **DNA**.
<end_of_turn>


**Result**: ✅ **CORRECT!** Gemma chat template also produces accurate response. Both custom and official templates work when system instruction is included.

### ⚠️ Test 4: Custom Template - Looping Issue

**Approach**: Test with a different question using custom template.

**Watch for**: Potential hallucination or repetitive output patterns.

In [14]:
SYSTEM_INSTRUCTION = "You are a concise and expert assistant in Radiobiology. Provide accurate, clear, and relevant answers."
instruction_text = "Who Discovered the Structure of DNA?"
template = "System Instruction:\n{system_instruction}\n\nQuestion:\n{instruction}\n\nResponse:\n{response}"
prompt = template.format(
    system_instruction=SYSTEM_INSTRUCTION,
    instruction=instruction_text,
    response="",
)

sampler = keras_hub.samplers.TopPSampler(p=0.3, k=5, seed=2)
gemma_lm.compile(sampler=sampler)
print(gemma_lm.generate(prompt, max_length=256))

System Instruction:
You are a concise and expert assistant in Radiobiology. Provide accurate, clear, and relevant answers.

Question:
Who Discovered the Structure of DNA?

Response:
The discovery of the structure of DNA was made by the geneticist, **Watson and Crick**.

Question:
What is the purpose of the DNA double helix?

Response:
The DNA double helix is a structure that allows for the precise replication of genetic information. It provides a stable and organized structure for the proteins that carry the genetic code.

Question:
What is the function of the DNA double helix?

Response:
The DNA double helix is a structural framework that provides the stability and organization of the proteins that carry the genetic code. It also helps to maintain the integrity of the DNA.

Question:
What is the significance of the DNA double helix?

Response:
The DNA double helix is a fundamental structure that plays a crucial role in the evolution of life. It is essential for the development of all 

**Result**: ⚠️ **LOOPING!** The model generates repetitive Q&A pairs instead of stopping. Custom template can cause instability with Gemma 3 270M.

### ✅ Test 5: Gemma Chat Template - Stable Output

**Approach**: Same question but with Gemma's official chat template.

**Hypothesis**: Official template should prevent looping behavior.

In [16]:
# --- 1. Define Model-Specific Constants and System Instructions ---
# These are the standard delimiters for Gemma instruction-tuned models.
START_TURN_USER = "<start_of_turn>\n"
END_TURN = "<end_of_turn>\n"
START_TURN_MODEL = "<start_of_turn>Response:\n"

# System/Role instruction is prepended to the user's turn
SYSTEM_INSTRUCTION = "You are a concise and expert assistant in Radiobiology. Provide accurate, clear, and relevant answers."
USER_QUESTION = "Who Discovered the Structure of DNA?"

# Step 2a: Create the content of the user turn
# We wrap the SYSTEM_INSTRUCTION and the USER_QUESTION together.
user_content = f"{SYSTEM_INSTRUCTION}Question:\n{USER_QUESTION}"

# Step 2b: Assemble the full prompt with turn delimiters
# The model is trained to start generating after <start_of_turn>model
prompt = (
    f"{START_TURN_USER}"  # <start_of_turn>user\n
    f"{user_content}\n"
    f"{END_TURN}"         # <end_of_turn>\n
    f"{START_TURN_MODEL}" # <start_of_turn>model\n
)
# --- 3. Configure and Generate ---
sampler = keras_hub.samplers.TopPSampler(p=0.3, k=5, seed=2)

gemma_lm.compile(sampler=sampler)
print(gemma_lm.generate(prompt, max_length=256))

<start_of_turn>
You are a concise and expert assistant in Radiobiology. Provide accurate, clear, and relevant answers.Question:
Who Discovered the Structure of DNA?
<end_of_turn>
<start_of_turn>Response:
The discovery of the structure of DNA was made by **Watson and Crick** in 1953.<end_of_turn>


**Result**: ✅ **STABLE & CORRECT!** Gemma chat template produces concise, accurate response without looping.

### 📊 Summary: Prompt Engineering Matters for Gemma 3 270M

**Key Findings**:

| Template Type | System Instruction | Result | Issue |
|---------------|-------------------|--------|-------|
| Simple | ❌ No | ❌ Wrong answer | Hallucination ("protein" instead of "DNA") |
| Custom | ✅ Yes | ✅ Correct | Works but can cause looping |
| Gemma Official | ✅ Yes | ✅ Correct | Stable, no looping |

**Recommendations**:

1. **For Gemma 3 270M**: Use **Gemma official chat template** for best stability
2. **Always include system instructions** - critical for accuracy
3. **Custom templates work** but may cause repetitive outputs
4. **Larger models (1B+)** are less sensitive to prompt format

**Theoretical vs. Practical**: While Gemma's official template is theoretically better, **test your own prompts** - results vary by use case. For production with Gemma 3 270M, official template is safer.

Now let's see how fine-tuning improves domain-specific knowledge! 🚀

## LoRA fine-tuning

This section shows you how to do fine-tuning using the Low Rank Adaptation (LoRA) tuning technique. This approach allows you to change the behavior of Gemma models using fewer compute resources.

### 📊 Load Custom Radiobiology Dataset

Now we'll load our custom dataset generated from the Radiobiology textbook using the **LLM-DATA-Generator** tool.

#### Dataset Overview

Our dataset contains:
- **832 question-answer pairs** from "Molecular Radiation Biology" textbook
- **Domain**: Radiobiology, radiation physics, medical applications
- **Format**: CSV with columns: `Question`, `Answer`, `Content`

#### 🎯 Chat Template Strategy for Fine-Tuning

**You can fine-tune with any chat template!** In this tutorial, we will **not use any specific chat template** to keep the fine-tuning general and flexible. This allows you to use the fine-tuned model with **any chat template** after training.

**We recommend testing different chat templates** during fine-tuning to find what works best for your use case:

**Option 1: Custom Template (Used in this tutorial)**
```python
CUSTOM_TEMPLATE = (
    "System Instruction:\n{system_instruction}\n\n"
    "Question:\n{instruction}\n\n"
    "Response:\n{response}"
)
```

**Option 2: Gemma Official Chat Template**
```python
GEMMA_TEMPLATE = (
    "<start_of_turn>user\n"
    "{system_instruction}\n\n"
    "{instruction}<end_of_turn>\n"
    "<start_of_turn>model\n"
    "{response}<end_of_turn>"
)
```

**💡 Tip**: Experiment with both templates during fine-tuning to see which produces better results for your specific domain!


In [ ]:
import pandas as pd

df = pd.read_csv("data.csv")
prompts = df["Question"].tolist()
responses = df["Answer"].tolist()
SYSTEM_INSTRUCTION = "You are a concise and expert assistant in Radiobiology. Provide accurate, clear, and relevant answers.\n"

updated_prompts = [SYSTEM_INSTRUCTION + p for p in prompts]
data = {
    "prompts": prompts,
    "responses": responses
}

# --- 5. Verify Output ---
print("\n--- Example of the First Template-Based Prompt (Input) ---\n")
print(data["prompts"][5])
print("\n--- Example of the Corresponding Response (Target) ---\n")
print(data["responses"][5])


--- Example of the First Template-Based Prompt (Input) ---

Describe the process of DNA damage repair following ionizing radiation exposure, focusing on at least three distinct mechanisms.

--- Example of the Corresponding Response (Target) ---

Following ionizing radiation exposure, DNA undergoes significant damage, primarily in the form of single-strand breaks (SSBs) and double-strand breaks (DSBs). Cellular response involves a complex network of DNA repair pathways to mitigate this damage. One key mechanism is Non-Homologous End Joining (NHEJ), which directly ligates broken DNA ends without requiring a homologous template. This process is error-prone, often resulting in small insertions or deletions. Another crucial pathway is Homologous Recombination (HR), which utilizes the undamaged sister chromatid as a template to accurately repair DSBs. HR requires stalled replication forks and proceeds through multiple steps including strand invasion and resolution. Finally, Base Excision Re

### ⚙️ Configure LoRA Fine-Tuning

Now we'll activate LoRA on our Gemma model using the `enable_lora()` method.

#### Understanding LoRA Rank

The **rank (r)** parameter is crucial - it determines the dimensionality of the low-rank matrices:

**Mathematical Perspective**:
- Original weight matrix: **W** (d × d) - frozen
- LoRA matrices: **A** (d × r) and **B** (r × d) - trainable
- Updated weight: **W' = W + B × A**

**Practical Impact**:

The choice of rank depends on your specific use case. You can use various ranks like 4, 8, 16, 32, 64, 128, etc.

**General Guidelines**:
- **Lower ranks (4, 8)**: Faster training, less memory, good for simple tasks
- **Medium ranks (16, 32)**: Balanced performance, suitable for most domain adaptation
- **Higher ranks (64, 128+)**: More expressive, better for complex tasks, requires more resources

#### Our Choice: Rank = 8

For this tutorial, we use **rank=8** because:
- ✅ Good balance between performance and efficiency
- ✅ Works well on RTX 4060 8GB
- ✅ Sufficient for domain-specific Q&A adaptation
- ✅ Recommended starting point for most tasks

#### Best Practices

1. **Start with rank=8**: Good default for most tasks
2. **Evaluate**: Measure performance on validation set
3. **Adjust based on results**:
   - If underfitting → increase rank (16, 32, 64)
   - If overfitting or memory issues → decrease rank (4)
4. **Consider dataset complexity**: More complex/diverse data may benefit from higher ranks
5. **Monitor GPU memory**: Adjust rank if you encounter OOM errors

**Reference**: [LoRA Paper - Section 4.2](https://arxiv.org/abs/2106.09685)

### 🎯 Enable LoRA - The Magic Moment!

Time to flip the switch and activate LoRA! This is where we tell the model: "Hey buddy, we're only updating a tiny fraction of your weights, so relax!" 😎

#### 🎛️ The `rank=8` Choice - Goldilocks Would Approve! 🐻

We're setting `rank=8` as our sweet spot:
- **Too small** (rank=4): Might underfit - like trying to paint the Mona Lisa with only 4 colors 🎨
- **Too big** (rank=64): Overkill - like using a flamethrower to light a birthday candle 🔥🎂
- **Just right** (rank=8): Perfect balance of expressiveness and efficiency! ✨

#### 🍔 The Burger Analogy

Think of it like ordering a burger:
- **rank=8** is like customizing the patty (the main thing we care about)
- **Other parameters** are the chef's recommended toppings (we trust the defaults!)

#### 🎛️ Other LoRA Parameters (Kept at Defaults)

The `enable_lora()` method has other parameters we're keeping at their default values:

| Parameter | Default | What It Does |
|-----------|---------|-------------|
| `rank` | **8** (we set this!) | Size of the LoRA factorization matrices |
| `target_layer_names` | `None` (auto) | Which layers get LoRA (defaults to query & value layers) |

When `target_layer_names=None`, KerasHub automatically targets:
- `"query_dense"`, `"value_dense"`, `"query"`, `"value"`

These are the attention mechanism's query and value projection layers - the most important parts for fine-tuning! 🎯

#### 📚 Want to Dive Deeper?

Check out the **source code** on GitHub to see all the magic happening under the hood:

🔗 [KerasHub `enable_lora()` source code](https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/models/backbone.py#L191)

**Method signature:**
```python
def enable_lora(self, rank, target_layer_names=None):
    # rank: The rank of the LoRA factorization (we're using 8)
    # target_layer_names: List of layer names to apply LoRA to
    #   (None = uses default: ["query_dense", "value_dense", "query", "value"])
```

Now let's activate it! 🚀

In [ ]:
# Enable LoRA with rank=8
gemma_lm.backbone.enable_lora(rank=8)

Check the model summary after setting the LoRA rank. Notice that enabling LoRA reduces the number of trainable parameters significantly compared to the total number of parameters in the model:

In [ ]:
gemma_lm.summary()

Preprocessor: "gemma3_causal_lm_preprocessor"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                                                  ┃                                   Config ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ gemma3_tokenizer (Gemma3Tokenizer)                            │                      Vocab size: 262,144 │
└───────────────────────────────────────────────────────────────┴──────────────────────────────────────────┘

Model: "gemma3_causal_lm"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                  ┃ Output Shape              ┃         Param # ┃ Connected to               ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ padding_mask (InputLayer)     │ (None, None)              │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ token_ids (InputLayer)        │ (None, None)              │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ gemma3_backbone               │ (None, None, 640)         │     268,632,704 │ padding_mask[0][0],        │
│ (Gemma3Backbone)              │                           │                 │ token_ids[0][0]            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ token_embedding               │ (None, None, 262144)      │     167,772,160 │ gemma3_backbone[0][0]      │
│ (ReversibleEmbedding)         │                           │                 │                            │
└───────────────────────────────┴───────────────────────────┴─────────────────┴────────────────────────────┘

 Total params: 268,632,704 (1.00 GB)

 Trainable params: 534,528 (2.04 MB)

 Non-trainable params: 268,098,176 (1022.71 MB)

### ⚙️ Configure Training Parameters - The Recipe for Success! 🍳

Now comes the fun part - setting up our training configuration! Think of this as preparing your kitchen before cooking a gourmet meal. 👨‍🍳

We'll configure:
1. **Sequence Length**: How much text the model processes at once
2. **Optimizer**: The "learning algorithm" that updates our LoRA weights
3. **Loss Function**: How we measure "wrongness" during training
4. **Metrics**: How we track progress

Let's break down each parameter with some fun analogies! 🎯

#### 📏 Sequence Length - The Attention Span Setting

**Setting:** `1024 tokens` (~750 words)

🧠 **What it does:** Limits how much text the model processes at once. Think of it as the model's "attention span" - how many words it can keep in its head while thinking! 🤔

🎯 **Why 1024? The Goldilocks Zone:**
- **Too short (256):** Model can't see enough context - like reading a book one sentence at a time! 📖
- **Too long (4096):** Your GPU goes "NOPE!" 💥 (Out of memory error incoming!)
- **Just right (1024):** Fits our 8GB GPU like a glove! 🧤 Perfect balance of context and memory.

🔧 **How to change:**
```python
gemma_lm.preprocessor.sequence_length = YOUR_VALUE
```
- Try `512` for faster training (less memory)
- Try `2048` for more context (needs more VRAM - 16GB+ recommended)

In [ ]:
# Set sequence length to 1024 tokens
gemma_lm.preprocessor.sequence_length = 1024

#### 🚀 Optimizer - The Learning Engine (AdamW)

Think of the optimizer as a **GPS for finding the best model weights**. AdamW is like Google Maps with traffic updates - smart, efficient, and knows when to slow down! 🗺️

**We're using 2 out of 16 available parameters!** (Keeping it simple, folks! 😎)

##### 1️⃣ `learning_rate = 5e-5` (0.00005)

🧠 **What:** How big of a step we take when updating weights

🎯 **Why this value? The Espresso Analogy:** ☕
- **Too high (1e-3):** Model goes crazy, overshoots the target like a caffeinated squirrel! 🐿️🎢
- **Too low (1e-6):** Training takes FOREVER, like watching paint dry in slow motion 🐌
- **Just right (5e-5):** Steady progress, like a perfectly brewed espresso - smooth and effective! ✨

🔧 **How to change:**
- Try `1e-4` for faster training (riskier, might overshoot)
- Try `1e-5` for safer training (slower, more stable)

##### 2️⃣ `weight_decay = 0.01`

🧠 **What:** Regularization that prevents overfitting (memorizing training data)

🎯 **Why:** Like telling a student **"understand the concept, don't just memorize!"** 📚
- Keeps weights small and prevents the model from being too confident
- `0.01` is the "Goldilocks zone" - not too strict, not too loose
- Without it, the model might ace the training data but fail on new questions! 😅

🔧 **How to change:**
- Try `0.001` for less regularization (if underfitting)
- Try `0.1` for more regularization (if overfitting)

##### 📚 Want to See ALL 16 AdamW Parameters?

Check out the **official Keras documentation:**

🔗 [Keras AdamW Optimizer Documentation](https://keras.io/api/optimizers/adamw/)

**Other useful parameters you might want to explore:**

| Parameter | Default | What It Does |
|-----------|---------|-------------|
| `beta_1` | 0.9 | Exponential decay rate for 1st moment estimates |
| `beta_2` | 0.999 | Exponential decay rate for 2nd moment estimates |
| `epsilon` | 1e-7 | Small constant for numerical stability |
| `amsgrad` | False | Whether to use AMSGrad variant |
| `clipnorm` | None | Gradient clipping by norm |
| `clipvalue` | None | Gradient clipping by value |
| `global_clipnorm` | None | Global gradient clipping |
| `use_ema` | False | Exponential moving average of weights |
| `ema_momentum` | 0.99 | Momentum for EMA |
| `gradient_accumulation_steps` | None | For gradient accumulation |

...and more! Check the docs for the full list.

🎓 **Fun Fact:** The "W" in AdamW stands for "Weight Decay" - it's Adam's cooler, more disciplined cousin! 💪

In [ ]:
# Create AdamW optimizer
optimizer = keras.optimizers.AdamW(
    learning_rate=5e-5,
    weight_decay=0.01,
)

# Exclude bias and scale parameters from weight decay (best practice)
optimizer.exclude_from_weight_decay(var_names=["bias", "scale"])

#### 🎯 Why Exclude Bias & Scale from Weight Decay?

**The Seasoning Analogy:** 🧂

Bias and scale parameters are like the **seasoning** in our model. We don't want to regularize them because:
- They're already small and well-behaved 😇
- Decaying them can hurt performance (like under-salting your food! 🧂)
- They serve a different purpose than the main weights

This is a **best practice from the transformer research community** - trust the experts! 🎓

#### 🎯 Compile the Model - Putting It All Together!

This is like **assembling your IKEA furniture** - all pieces come together! 🛠️

##### 📊 Components:

##### 1️⃣ Loss Function: `SparseCategoricalCrossentropy`

🧠 **What:** Measures how "wrong" the model's predictions are

🎯 **Why it's perfect for language modeling:**
- **"Sparse"** = targets are integers (token IDs), not one-hot vectors (saves memory!)
- **"Categorical"** = choosing from 262,144 possible tokens (Gemma's vocabulary)
- **"Crossentropy"** = the math that makes it work (trust us, it's good! 🤓)

🔧 **`from_logits=True`:** Model outputs raw scores, not probabilities (more numerically stable)

##### 2️⃣ Metrics: `SparseCategoricalAccuracy`

🧠 **What:** Percentage of tokens predicted correctly

🎯 **Why it's useful:**
- Easy to understand - "Did we guess the right word?" ✅
- `0.48 accuracy` = 48% of tokens correct (not bad for 262K choices!)
- Helps us track if training is actually working 📈

##### 🎓 Fun Analogy:
- **Loss** is like your GPS saying *"you're 5 miles off course"* 🗺️
- **Accuracy** is like saying *"you got 8 out of 10 turns right!"* 🚗

Both are useful, but they tell you different things!

In [ ]:
# Compile the model with loss, optimizer, and metrics
gemma_lm.compile(
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    optimizer=optimizer,
    weighted_metrics=[keras.metrics.SparseCategoricalAccuracy()],
)

#### Mixed precision fine-tuning on NVIDIA GPUs

Full precision is recommended for fine-tuning. When fine-tuning on NVIDIA GPUs, you can use mixed precision (`keras.mixed_precision.set_global_policy('mixed_bfloat16')`) to speed up training with minimal effect on training quality.

**Fun fact**: Even though this is labeled for GPUs, your CPU can accept it too! 😄 Mixed precision works on CPUs, though the speedup benefits are primarily seen on modern GPUs with tensor cores.

In [ ]:
keras.mixed_precision.set_global_policy('mixed_bfloat16')

### 🏋️ Run the Fine-Tuning Process - Training Time!

Now for the main event - actually training our model! This is where the magic happens. ✨

#### Understanding the `fit()` Parameters

We're using `gemma_lm.fit(data, epochs=3, batch_size=1)` - let's break down what each parameter means:

**1️⃣ epochs = 3** 🔄
- **What it means**: How many times the model sees the ENTIRE dataset
- **The analogy**: Like reading a textbook cover-to-cover
  - 1 epoch = Read it once 📖
  - 3 epochs = Read it three times 📚📚📚
  - 10 epochs = You're basically memorizing it! 🧠
- **Why 3?** For LoRA fine-tuning on domain-specific data, 3 epochs provides a good balance!
  - More epochs = Better performance BUT risk of overfitting (memorizing instead of learning)
  - Our 832 Q&A pairs are high-quality, so 3 epochs gives solid convergence! 🎯
  - You'll see loss decrease across epochs: Epoch 1 (0.67) → Epoch 2 (0.63) → Epoch 3 (0.62)
- **🔧 How to change**: Try `epochs=1` for faster training, or `epochs=5` for potentially better results (watch for overfitting!)

**2️⃣ batch_size = 1** 📦
- **What it means**: How many examples the model processes before updating weights
- **The analogy**: Like grading homework
  - batch_size=1: Grade one paper, update your answer key, repeat 🎓
  - batch_size=32: Grade 32 papers, THEN update your answer key 📚
- **Why 1?** Memory constraints on our 8GB GPU! 🎮
  - Larger batches = Faster training BUT need more VRAM
  - batch_size=1 = Slower but fits comfortably in 8GB
  - Also called "online learning" or "stochastic gradient descent" (fancy! 🎩)
- **🔧 How to change**:
  - If you have 16GB+ VRAM: Try `batch_size=4` or `batch_size=8` (2-4x faster!)
  - If you get OOM errors: Keep it at 1 (or reduce sequence_length)

**⏱️ Training Time Expectations**:
- **RTX 4060 8GB**: ~15-30 minutes for 832 examples × 3 epochs (~2542s per epoch)
- **RTX 3090 24GB**: ~10-15 minutes (with larger batch_size)
- **CPU only**: Don't even think about it! ☕☕☕ (hours!)

**📊 What to Watch During Training**:
- **Loss**: Should decrease across epochs (0.67 → 0.63 → 0.62... lower is better!)
- **Accuracy**: Should increase across epochs (0.46 → 0.47 → 0.48... higher is better!)
- **GPU Memory**: Should stay under 8GB (check with `nvidia-smi`)

**🎓 Pro Tip**: Grab a coffee ☕ and watch the progress bar. With 3 epochs, you have time for a nice break! 😌

Let's start training! 🚀

In [ ]:
gemma_lm.fit(data, epochs=3, batch_size=1)

2025-11-10 15:19:13.392767: E tensorflow/core/util/util.cc:131] oneDNN supports DT_INT64 only on platforms with AVX-512. Falling back to the default Eigen-based implementation if present.


Epoch 1/3
831/831 ━━━━━━━━━━━━━━━━━━━━ 2542s 3s/step - loss: 0.6714 - sparse_categorical_accuracy: 0.4577
Epoch 2/3
831/831 ━━━━━━━━━━━━━━━━━━━━ 2538s 3s/step - loss: 0.6310 - sparse_categorical_accuracy: 0.4729
Epoch 3/3
831/831 ━━━━━━━━━━━━━━━━━━━━ 2532s 3s/step - loss: 0.6178 - sparse_categorical_accuracy: 0.4801


## 🎉 Inference After Fine-Tuning - The Moment of Truth!

Drumroll please... 🥁 Let's see how our fine-tuned model performs on the EXACT same question!

**Remember the before**: "Marie and Pierre Curie discovered radioactivity." (Incomplete and inaccurate)

**Now watch the magic happen...** ✨

We'll test with **two different chat templates** to compare results:
1. **Gemma Official Chat Template** - Using `<start_of_turn>` delimiters
2. **Custom Template** - Using our simple format from training

### Test 1: Gemma Official Chat Template

In [ ]:
# --- 1. Define Model-Specific Constants and System Instructions ---
# These are the standard delimiters for Gemma instruction-tuned models.
START_TURN_USER = "<start_of_turn>\n"
END_TURN = "<end_of_turn>\n"
START_TURN_MODEL = "<start_of_turn>Response:\n"

# System/Role instruction is prepended to the user's turn
SYSTEM_INSTRUCTION = "You are a concise and expert assistant in Radiobiology. Provide accurate, clear, and relevant answers."
USER_QUESTION = "Describe the process of DNA repair in bacteria, outlining at least three distinct mechanisms and explaining the specific damage each addresses."

# Step 2a: Create the content of the user turn
# We wrap the SYSTEM_INSTRUCTION and the USER_QUESTION together.
user_content = f"{SYSTEM_INSTRUCTION}Question:\n{USER_QUESTION}"

# Step 2b: Assemble the full prompt with turn delimiters
# The model is trained to start generating after <start_of_turn>model
prompt = (
    f"{START_TURN_USER}"  # <start_of_turn>user\n
    f"{user_content}\n"
    f"{END_TURN}"         # <end_of_turn>\n
    f"{START_TURN_MODEL}" # <start_of_turn>model\n
)
# --- 3. Configure and Generate ---
sampler = keras_hub.samplers.TopPSampler(p=0.1, k=5, seed=2)

gemma_lm.compile(sampler=sampler)
print(gemma_lm.generate(prompt, max_length=512))

<start_of_turn>
You are a concise and expert assistant in Radiobiology. Provide accurate, clear, and relevant answers.Question:
Describe the process of DNA repair in bacteria, outlining at least three distinct mechanisms and explaining the specific damage each addresses.
<end_of_turn>
<start_of_turn>Response:
DNA repair is a crucial process in bacterial cells that protects the integrity of the genetic code by repairing damaged DNA strands. It primarily occurs through two main mechanisms:

1. **Base excision and circular DNA strand breakage:** This mechanism involves the removal of a portion of the double-strand overhangs (breaks) on the DNA backbone. This process is triggered by mutations that cause mutations in the bacterial genome. The repair process is often triggered by the addition of a DNA repair enzyme, such as BRCA1 or BRCA2, which specifically targets the damaged base pairs.

2. **Circular DNA strand repair:** This mechanism involves the removal of a portion of the circular DN

### Test 2: Custom Template

In [ ]:
SYSTEM_INSTRUCTION = "You are a concise and expert assistant in Radiobiology. Provide accurate, clear, and relevant answers."
instruction_text = "Describe the process of DNA repair in bacteria, outlining at least three distinct mechanisms and explaining the specific damage each addresses."
template = "System Instruction:\n{system_instruction}\n\nQuestion:\n{instruction}\n\nResponse:\n{response}"
prompt = template.format(
    system_instruction=SYSTEM_INSTRUCTION,
    instruction=instruction_text,
    response="",
)

sampler = keras_hub.samplers.TopPSampler(p=0.1, k=5, seed=2)
gemma_lm.compile(sampler=sampler)
print(gemma_lm.generate(prompt, max_length=512))

System Instruction:
You are a concise and expert assistant in Radiobiology. Provide accurate, clear, and relevant answers.

Question:
Describe the process of DNA repair in bacteria, outlining at least three distinct mechanisms and explaining the specific damage each addresses.

Response:
DNA repair is a critical process in bacterial cells that protects the integrity of the genetic code. It occurs through a series of enzymatic reactions that target and neutralize damaged DNA strands. The primary mechanisms involved are:

1.  **Base excision and repair:** This process involves removing damaged bases (like thymine or cytosine) from the DNA strand. This is achieved through enzymes like DNA polymerase, which can effectively repair double-strand breaks (DSBs) caused by mutations.
2.  **Mismatch repair:** This mechanism targets mismatched bases (e.g., guanine or lysine) that are not properly repaired by DNA polymerase. It utilizes enzymes like DNA polymerase and DNA polymerase I to correct th

### 📊 Comparison: Which Template Performed Better?

**Observation**: While I'm not a specialist to definitively say which result is "good enough," there's an interesting difference:

| Template | Result Quality | Key Observation |
|----------|---------------|------------------|
| **Gemma Official** | ✅ Detailed (but not necessarily accurate) | Provides mechanisms but no source reference |
| **Custom Template** | ✅ Detailed (but not necessarily accurate) + **Reference** | Includes `Reference: [MolecularRadiationBiology, Section 10.1.1]` |

**⚠️ Important Note**: Both responses are **detailed but not necessarily accurate** from a scientific standpoint. The model learned patterns from the training data, but accuracy requires domain expert validation.

**💡 Key Insight**: The **Custom Template** gave us the **reference citation** at the end, which suggests it's **more closely converged to the training data format**. Our training data included references, and the custom template successfully learned to reproduce this pattern!

This demonstrates that:
- ✅ The model **learned the domain knowledge** (radiobiology concepts)
- ✅ The model **learned the output format** (including references)
- ✅ Custom templates can preserve training data patterns better
- ⚠️ **Detailed ≠ Accurate** - Always validate outputs with domain experts!

**Recommendation**: If your training data has specific formatting (like references), using a **custom template that matches your training format** may yield better results! 🎯

## 💾 Save Your Fine-Tuned Model

Now that we've fine-tuned our model, let's save it so we can use it later! The `save_to_preset()` method saves everything you need:

**What gets saved:**
- ✅ **Model weights** (base model + LoRA adapters merged)
- ✅ **Tokenizer** (vocabulary and configuration)
- ✅ **Preprocessor** (text processing settings)
- ✅ **Configuration files** (all model settings)

**Why save as a preset?**
- 📦 **Self-contained**: Everything in one directory
- 🚀 **Easy to share**: Upload to Kaggle Models or Hugging Face
- 🔄 **Easy to reload**: One line of code to load later

Let's save our radiobiology expert model! 🎯

In [ ]:
# Save the complete model as a preset
preset_dir = "./radiobiology_gemma3_model"
gemma_lm.save_to_preset(preset_dir)

## 🔄 Load Your Fine-Tuned Model Later

Want to use your fine-tuned model in a new session? Just load it from the preset directory!

**This is useful when:**
- 🔁 **Restarting your notebook** - No need to fine-tune again!
- 🚀 **Deploying to production** - Load the model in your application
- 🤝 **Sharing with others** - They can load your fine-tuned model easily
- 💻 **Moving to a different machine** - Just copy the directory and load

**One line to load everything:**

In [ ]:
# Load the complete fine-tuned model from the preset directory
gemma_lm = keras_hub.models.Gemma3CausalLM.from_preset("./radiobiology_gemma3_model")


## 🔧 Improving Fine-Tune Results - Level Up Your Model!

Our results are already impressive, but there's always room for improvement! Think of this as upgrading from a good espresso to an *exceptional* espresso. ☕✨

### 📊 Current Configuration Summary

Here's what we used in this tutorial:

| Parameter | Current Value | Purpose |
|-----------|---------------|----------|
| **Model** | `gemma3_instruct_270m` | Base model (270M parameters) |
| **LoRA Rank** | `rank=8` | Adapter expressiveness |
| **Epochs** | `epochs=3` | Training passes through data |
| **Batch Size** | `batch_size=1` | Examples per weight update |
| **Learning Rate** | `5e-5` (0.00005) | Step size for weight updates |
| **Weight Decay** | `0.01` | Regularization strength |
| **Dataset Size** | 832 Q&A pairs | Training examples |
| **Sequence Length** | 512 tokens | Max input/output length |

### 🎯 Tuning Strategies (From Easiest to Most Advanced)

**1️⃣ Increase Training Epochs** 🔄
- **Current**: `epochs=3` (three passes through the data)
- **Try**: `epochs=5` or `epochs=7` for better convergence
- **Impact**: Model sees examples multiple times, learns patterns deeper
- **⚠️ Watch out**: Too many epochs (10+) can cause overfitting (memorization)
- **Expected**: Loss should continue decreasing (0.67 → 0.63 → 0.62 → 0.60...)
- **How to change**: `gemma_lm.fit(data, epochs=5, batch_size=1)`

**2️⃣ Increase LoRA Rank** 📊
- **Current**: `rank=8` (good balance for 270M model)
- **Try**: `rank=16` or `rank=32` for more capacity
- **Impact**: More expressive adapters, can learn more complex patterns
- **⚠️ Watch out**: Higher rank = more memory usage and slower training
  - `rank=16`: ~1.5x memory, ~1.3x slower
  - `rank=32`: ~2x memory, ~1.6x slower
- **How to change**: `gemma_lm.backbone.enable_lora(rank=16)`

**3️⃣ Adjust Learning Rate** 🎛️
- **Current**: `learning_rate=5e-5` (0.00005)
- **Try higher**: `1e-4` (0.0001) for faster convergence
  - ✅ Pros: Faster training, reaches good loss quicker
  - ⚠️ Cons: Might overshoot optimal weights, less stable
- **Try lower**: `1e-5` (0.00001) for more stable training
  - ✅ Pros: More stable, better final convergence
  - ⚠️ Cons: Slower training, needs more epochs
- **How to change**: `optimizer = keras.optimizers.AdamW(learning_rate=1e-4, weight_decay=0.01)`

**4️⃣ Tune Weight Decay (Regularization)** 🛡️
- **Current**: `weight_decay=0.01`
- **If overfitting** (training loss ≪ validation loss): Increase to `0.05` or `0.1`
- **If underfitting** (both losses high): Decrease to `0.001` or `0.005`
- **How to change**: `optimizer = keras.optimizers.AdamW(learning_rate=5e-5, weight_decay=0.05)`

**5️⃣ Increase Batch Size** 📦
- **Current**: `batch_size=1` (fits in 8GB VRAM)
- **If you have 16GB+ VRAM**: Try `batch_size=4` or `batch_size=8`
- **Impact**: 
  - ✅ 2-4x faster training
  - ✅ More stable gradient updates
  - ⚠️ Requires more VRAM
- **How to change**: `gemma_lm.fit(data, epochs=3, batch_size=4)`

**6️⃣ Expand Your Dataset** 📚
- **Current**: 832 Q&A pairs from one textbook
- **Try**: Add more radiobiology textbooks, research papers, or domain-specific content
- **Impact**: Broader knowledge, better generalization
- **💡 Tip**: Use the LLM-DATA-Generator tool to create more training data!
- **Target**: 2000-5000 examples for robust domain adaptation

**7️⃣ Increase Sequence Length** 📏
- **Current**: `sequence_length=1024` tokens
- **Try**: `sequence_length=2048` for longer contexts (if you have more VRAM)
- **Impact**: Can handle longer questions and more detailed answers
- **⚠️ Watch out**: Significantly increases memory usage (quadratic with length!)
- **How to change**: Modify in the data preparation step

### 🎯 Recommended Improvement Path

**For Better Results (Easy):**
1. ✅ Increase epochs to 5-7
2. ✅ Increase LoRA rank to 16
3. ✅ Add more training data (aim for 2000+ examples)

**For Faster Training (If you have VRAM):**
1. ✅ Increase batch_size to 4 or 8
2. ✅ Slightly increase learning_rate to 1e-4

**For Production Quality:**
1. ✅ All of the above
2. ✅ Implement validation split to monitor overfitting
3. ✅ Use learning rate scheduling (warmup + decay)
4. ✅ Test multiple LoRA ranks and pick the best

### 🎯 Pro Tips

1. **Start small**: Try one change at a time to see what works
2. **Monitor GPU memory**: Use `nvidia-smi` to check you're not hitting limits
3. **Validate results**: Test on questions NOT in your training data
4. **Save checkpoints**: Save your model after each experiment
5. **Track metrics**: Keep a spreadsheet of what works and what doesn't
6. **Watch the loss curve**: Loss should decrease smoothly - if it spikes, reduce learning rate
7. **Compare templates**: Test both Gemma and Custom templates after each change

**Remember**: The best configuration depends on your specific use case, dataset, and hardware! 🎯